ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
-----
-__kernel void __khr_io_stream_in_uchar(
-   uchar *output, size_t *num) {
-   // It is not feasible to describe this kernel in OpenCL C as I/O devices
-   // are not representable with it.
-}
-----
+* `rankof(A) == rankof(B) >= 2`.
+* Let `shapeof(A~t~) == (b..., m, k)` and `shapeof(B~t~) = (b..., k,
+  n)` of tensors `A` and `B`, respectively, after possible tranposing.
+  `shapeof\(R)` must be `(b..., m, n)`.
+* `elementof(A) == elementof(B)`
+* `elemkindof\(R) == elemkindof(A)`
+* `elementof\(R) == elementof(A)` or `elementof(A)` is promotable to
+  `elementof\(R)` without loss of meaning.
+// E.g. cl_int -> cl_uint: loses negative values
+| *Kernel mode properties*
+a|
+This DBK accepts the following properties:
 
-| *khr_io_stream_out_uchar*
-| Non-blocking write of data to an output/sink associated with the device.
-| -
+* CL_DBK_PROPERTY_MAX_RELATIVE_ERROR: Unset property defaults to positive infinity.
+|
+| Name: *khr_leaky_relu_v1*
+| *Kernel Attributes*
 a|
-[start=0]
-. uchar* [in]: The data to write.
-. size_t* [in+out]: In: Number of bytes to write. Out: Number of bytes that could be written (can be 0).
-4+| OpenCL C Semantics
-4+a|
-[source,c]
-----
-__kernel void __khr_io_stream_out_uchar(
-   uchar *input, size_t *num) {
-   // It is not feasible to describe this kernel in OpenCL C as I/O devices
-   // are not representable with it.
-}
-----
+Fields of the `cl_dbk_leaky_relu_v1` structure:
+. cl_tensor_desc_t in: Input tensor description.
+. cl_tensor_desc_t out: Output tensor description.
+. cl_float alpha: Coefficient of leakage.
+| *Kernel arguments*
+a|
+. cl_tensor_t in: The input tensor.
+. cl_tensor_t out: The output tensor.
+| *Description*
+a|
+Applies operation `alpha * x if x < 0 else x` on all
+elements of the `in` tensor.
+
+If target device does not support denormals, then `alpha` is flushed
+to zero before the operation is applied.
 
-| *khr_io_stream_in_blocking_uchar*
-| Blocking read of data from a sensor/stream associated with the device.
-a| -
+| *Kernel mode properties*
+| N/A
+| *Attribute validation rules*
 a|
-[start=0]
-. uchar* [out]: The data.
-* size_t* [in]: How many bytes to read before returning.
+* `shapeof(in) == shapeof(out)`
+* `elementof(in) == elementof(out)`
+* `alpha` must be a finite value.
+|===
+
+==== Launching DBKs from the Device Side ====
+
+DBKs are primarily meant to be launched as kernel commands via
+host-side command queues.  Optionally, they can be callable from
+device-side via `enqueue_kernel`:
+
+TBC. This probably needs device-side function corresponding to
+clCreateDefinedBuiltInKernelDescriptor.
+
+==== Sample Code ====
 
-4+| OpenCL C Semantics
-4+a|
 [source,c]
 ----
-__kernel void __khr_io_stream_in_blocking_uchar(uchar *output, size_t *num) {
-   while (*num) {
-       size_t num_read = *num;
-       __khr_io_stream_in_uchar(output, &num_read);
-       num -= num_read;
-       output += num_read;
-   }
+// TBD. Similarly in cl_qcom_ml_ops, tensors have type
```

Comment:
Perhaps just kick off the cl_tensor extension as another PR to my repo where we can discuss its contents separately. Also we should have examples of DBKs which do not use cl_tensor but cl_mem or even scalar args.

Reply:
> Also we should have examples of DBKs which do not use cl_tensor but cl_mem or even scalar args. 

What would be the DBK that wouldn't use `cl_tensor`s and why?